### DISCUSSION TITLE
Compared with version 7.0.6.9, the encoding delay of gif format in version 7.1.0-52 has increased several times

### DISCUSSION THREAD

**Discussion by zhouleiMax** [ImageMagick/ImageMagick#6454] [Help]
### ImageMagick version

7.1.0-52

### Operating system, version and so on

Linux VM-232-130-centos 3.10.107-1-tlinux2_kvm_guest-0054

### Description

Compared with version 7.0.6.9, the encoding delay of gif format in version 7.1.0-52 has increased several times
My test sample is a 180 frame picture，
7.1.0-52 ：save:158846.52ms
7.0.6.9：   save:13614.06ms

**Comment by zhouleiMax** [comment]
[52.zip](https://github.com/ImageMagick/ImageMagick/files/11936777/52.zip)
it's my test sample

**Comment by zhouleiMax** [comment]
I compared the difference between the two versions in the gif encoding process and found that the time is mainly consumed in the CompareImagesBounds() function（MagcikCore/layer.c）

**Comment by zhouleiMax** [comment]
Before encoding the gif format, I used the OptimizeImageLayers() function

**Comment by fmw42** [comment]
It might help the developers if you show your exact commands and provide your input images

**Reply by zhouleiMax** [reply]
Hello, I am also a developer. Using the IM library (libMagickCore-7.Q8.a), I started with version 7.0.6-9 and recently updated to version 7.1.0-52. I found that the time when encoding gif becomes It is very large, and it has been doubled. In my code, the OptimizeImageLayers() function is used. By recording the time, I found that the time consumption of this function in the two versions is very different. Then I checked the source code and found that the time consumption is in CompareImagesBounds ()function

**Reply by zhouleiMax** [reply]
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <string>
#include <time.h>
#include <MagickCore.h>

#include <iostream>
#include <opencv2/opencv.hpp>
#include <opencv2/core/core.hpp>
#include <opencv2/core/core_c.h>
#include <opencv2/imgproc/imgproc.hpp>
#include <opencv2/highgui/highgui.hpp>
#include <opencv2/imgproc/imgproc_c.h>
#include <opencv2/imgproc/types_c.h>
#include <highgui.hpp>

using namespace cv;

static bool read_file(const char* filepath, std::string& data)
{
    FILE* fp = fopen(filepath, "rb");
    if (!fp) {
        fprintf(stderr, "read_file %s failed: %d %s\n", filepath, errno, strerror(errno));
        return false;
    }
    fseek(fp, 0, SEEK_END);
    int len = ftell(fp);
    rewind(fp);

    char* buffer = new char[len + 1];
    fread(buffer, 1, len, fp);
    fclose(fp);

    buffer[len] = '\0';
    data.resize(len);
    copy(buffer, buffer + len, data.begin());
    delete[] buffer;

    return true;
}

int main(int argc,char **argv)
{
  const char* filepath = "F://gif//52.gif";
  const unsigned char *header;
  int src_size;
  std::string filedata;

  read_file(filepath, filedata);
  header = (const unsigned char *)filedata.data();  
  src_size = filedata.size();
  printf("src_size = %d \n", src_size);
  
  MagickCoreGenesis(NULL, MagickFalse); // Initialize the Magick environment.
  ImageInfo* image_info = AcquireImageInfo();
  ExceptionInfo* exception = AcquireExceptionInfo();
  Image* m_image = BlobToImage(image_info, header, src_size, exception);
  
  int m_needOptimize = 0; 
  Image* mm = m_image;
  int gifwidth = m_image->page.width != 0 ? m_image->page.width : m_image->columns;
  int gifheight = m_image->page.height != 0 ? m_image->page.height : m_image->rows;
  // Determine whether to optimize
  while (mm != NULL) {
    if(mm->columns < (gifwidth - 10) &&  mm->rows < (gifheight - 10))
    {
      m_needOptimize = 1;
      break;
    }
    else
      m_needOptimize = 0;
      mm = mm->next;
    } 
  Image* coalesced = CoalesceImages(m_image, exception); 
  m_image = DestroyImageList(m_image);
  m_image = coalesced;
  coalesced = NULL;

  Image * layers;
  clock_t c1 = clock();
  if(m_needOptimize){
    Image * layers = OptimizeImageLayers(m_image,exception);
  }
  clock_t c2 = clock();
  clock_t r = c2 - c1;
  printf("time-gap = %d \n", r);
  
  exception=DestroyExceptionInfo(exception);
  image_info=DestroyImageInfo(image_info);
  m_image=DestroyImageList(m_image);
  layers=DestroyImageList(layers);
  
  MagickCoreTerminus(); // destroys the MagickCore environment.

  waitKey(0);
  system("pause");
  return(0);
}

**Reply by zhouleiMax** [reply]
printf("time-gap = %d \n", r); = 60ms， it's too large ,The previous version was only ten milliseconds

**Reply by zhouleiMax** [reply]
![企业微信截图_16884421143879](https://github.com/ImageMagick/ImageMagick/assets/103361315/b8ba665c-0804-4298-b88a-11027dba8c46)

**Comment by zhouleiMax** [comment]
The delay has increased a lot, have you reproduced it?

**Comment by dlemstra** [comment]
We have seen this discussion but haven't had time to investigate this. This is a project that we work on in our spare time so we have no idea when we will be able to work on this. I could be weeks or months before we will be able to look at this.

**Comment by fmw42** [comment]
I suggest you test similar commands in command line mode and see if that has the same issue. If not, then it is in your code, not Imagemagick.

**Reply by zhouleiMax** [reply]
The purpose of using the function OptimizeImageLayers() is to make the encoding file as small as possible. The two versions found that the time spent here is very different. I saw that the underlying source code has been changed a lot. You can try this picture I uploaded. After decoding, run the OptimizeImageLayers function to test the time-consuming. I also tested it with the command line, but the command line does not necessarily use the OptimizeImageLayers() function, so it cannot be compared. You can write a script locally to decode the image, for example so:
Image * layers;
clock_t c1 = clock();
Image * layers = OptimizeImageLayers(m_image,exception);
clock_t c2 = clock();
clock_t r = c2 - c1;
printf("time-gap = %d \n", r);
[Uploading 52.zip…]()

**Reply by fmw42** [reply]
Add -layers optimize to the command line

`magick -dispose XXX image1 image2 ... imageN -loop 0 -layers optimize result.gif`

**Reply by zhouleiMax** [reply]
![企业微信截图_16886976482458](https://github.com/ImageMagick/ImageMagick/assets/103361315/bf3a9a58-b39c-4ee3-b96b-63da21c054e4)
![企业微信截图_16886977756538](https://github.com/ImageMagick/ImageMagick/assets/103361315/95de35f4-587d-4340-9efd-af13c69943c9)

**Reply by zhouleiMax** [reply]
I compiled these two versions, the same command test, the encoding delay is as above

**Reply by fmw42** [reply]
Please don't post images of code or messages

**Comment by urban-warrior** [comment]
The gold standard for any ImageMagick version 7 image processing method is ImageMagick version 6.  We timed `-layers optimization` for version 6 & 7 and got:

```
6.9.12-91: 0:55.96
7.1.1-13: 1:07.50
````

The few seconds difference can be accounted for by version 6 using `unsigned shorts` and version 7, `floats`.  Floating point ops are generally slower than unsigned ops.  In addition, ImageMagick version 7 has multi-spectral support.   Indexing across variable channels does diminish performance slightly as compared to the fixed pixel packets in version 6.

ImageMagick 7.0.6-10 performance is 0:09.40 suggesting a bug in that release that was subsequently fixed.  And although 7.0.6 appears to produce expected results for your use case, it clearly skips steps that are needed for a robust error free implementation of the optimization algorithm to achieve that performance level.

We could assume both modern ImageMagick 6 and 7 have a similar bug that affects performance, however, if we trace revision history, which would be quite time consuming, its likely we would stumble on a previous bug fix that subsequently slowed the performance to ensure a robust implementation.
